File: quick_sort.ml
let rec quick_sort lst = match lst with
    [] -> []
  | pivot :: rest ->
     let smaller_equal = List.filter (fun x -> x <= pivot) rest in
     let greater = List.filter (fun x -> pivot < x) rest in
     (quick_sort smaller_equal) @ [pivot] @ (quick_sort greater)


let rec is_sorted lst pre = match lst with
    [] -> true
  | first :: rest ->
     if first <= pre then is_sorted first rest
     else false

File: sugoiHaskell/quickSort.hs
quickSort :: (Ord a) => [a] -> [a]
quickSort [] = []
quickSort (x:xs) =
  let smallerEqual = [a | a <- xs, a <= x]
      larger = [a | a <- xs, x < a] in 
    quickSort smallerEqual ++ [x] ++ quickSort larger


isSorted :: (Ord a) => [a] -> Bool
isSorted [] = True
isSorted [x] = True
isSorted (x:y:xs)
  | x <= y = isSorted (y:xs)
  | otherwise = False
